### ISSUE TITLE
add `volatile` keyword in asm to suppress the compiler's optimisations

### ISSUE THREAD

**Issue by axiqia** [ashwio/arm64-sysreg-lib#1]
I wrote a simple driver to modify the system register.  It includes a header file in `arm64-sysreg-lib/include/sysreg/pmcr_el0.h`. 
``` c
        printk(KERN_ALERT "raw variable value: %llx\n", pmcr_val._);
        pmcr_val.e = 1;
        printk(KERN_ALERT "after modifid variable value: %llx\n", pmcr_val._);
        unsafe_write_pmcr_el0(pmcr_val); 
        pmcr_val = read_pmcr_el0();
        printk(KERN_ALERT "read from pmcr_val reg: %llx\n", pmcr_val._);
```

It seems that the output is not as expected. The output is like below: 
```
[351940.892080] raw variable value: 48006040
[351940.892398] after modifid variable value: 48006041
[351940.892877] read from pmcr_val reg: 48006040
```

The initial value of variable `pmcr_val`  is `48006040` and we modified it to  `48006041`. Then, I wrote it into the register and read it out. The current value of the variable should be `48006041`, but it is actually `48006040` as the output.

I intermixed the source code with disassembly.
``` 
       pmcr_val.e = 1;
  9c:   b2400275        orr     x21, x19, #0x1
        printk(KERN_ALERT "raw variable value: %llx\n", pmcr_val._);
  a0:   aa1303e1        mov     x1, x19
  a4:   90000000        adrp    x0, 0 <init_module>
  a8:   91000000        add     x0, x0, #0x0
  ac:   94000000        bl      0 <printk>
        printk(KERN_ALERT "after modifid variable value: %llx\n", pmcr_val._);
  b0:   aa1503e1        mov     x1, x21
  b4:   90000000        adrp    x0, 0 <init_module>
  b8:   91000000        add     x0, x0, #0x0
  bc:   94000000        bl      0 <printk>
  c0:   d51b9c15        msr     pmcr_el0, x21
        printk(KERN_ALERT "read from pmcr_val reg: %llx\n", pmcr_val._);
  c4:   90000000        adrp    x0, 0 <init_module>
  c8:   aa1303e1        mov     x1, x19
  cc:   91000000        add     x0, x0, #0x0
  d0:   94000000        bl      0 <printk>
```
It seems that the compiler deleted the read expression.

As a contrast, I added the `volatile` keyword to read and write the register. 
```
        printk(KERN_ALERT "raw variable value: %llx\n", pmcr_val._);
        pmcr_val.e = 1;
        printk(KERN_ALERT "after modifid variable value: %llx\n", pmcr_val._);
        __asm volatile(
            "MSR s3_3_c9_c12_0, %0"
            : /**/
            : "r"(pmcr_val._));

        __asm volatile(
            "MRS %0, s3_3_c9_c12_0"
            : "=r"(pmcr_val._));
        printk(KERN_ALERT "read from pmcr_val reg: %llx\n", pmcr_val._);
```
Below is the assembly code 
```
        pmcr_val.e = 1;
  38:   b2400293        orr     x19, x20, #0x1
        printk(KERN_ALERT "raw variable value: %llx\n", pmcr_val._);
  3c:   aa1403e1        mov     x1, x20
  40:   90000000        adrp    x0, 0 <init_module>
  44:   91000000        add     x0, x0, #0x0
  48:   94000000        bl      0 <printk>
        printk(KERN_ALERT "after modifid variable value: %llx\n", pmcr_val._);
  4c:   aa1303e1        mov     x1, x19
  50:   90000000        adrp    x0, 0 <init_module>
  54:   91000000        add     x0, x0, #0x0
  58:   94000000        bl      0 <printk>
        __asm volatile(
  5c:   d51b9c13        msr     pmcr_el0, x19
        __asm volatile(
  60:   d53b9c01        mrs     x1, pmcr_el0
        printk(KERN_ALERT "read from pmcr_val reg: %llx\n", pmcr_val._);
  64:   90000000        adrp    x0, 0 <init_module>
```
In short, we should add the `volatile` keyword when reading and writing registers to prevent the compiler from deleting the expression.

**Comment by ashwio** [comment]
Good catch! Yep we absolutely need the 'volatile' there. I've merged your patch. Thanks!
